segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java

```diff
@@ -0,0 +1,222 @@
+     * The transaction commit operation fails if this operation fails.
+     */
+    @Getter
+    @Setter
+    private Callable<Void> externalCommitStep;
+
+    /**
+     * Constructor.
+     *
+     * @param store   Underlying metadata store.
+     * @param version Version number of the transactions.
+     */
+    public MetadataTransaction(ChunkMetadataStore store, long version) {
+        this.store = Preconditions.checkNotNull(store, "store");
+        this.version = version;
+        data = new ConcurrentHashMap<String, BaseMetadataStore.TransactionData>();
+    }
+
+    /**
+     * Retrieves the metadata for given key.
+     *
+     * @param key key to use to retrieve metadata.
+     * @return Metadata for given key. Null if key was not found.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public StorageMetadata get(String key) throws StorageMetadataException {
+        return store.get(this, key);
+    }
+
+    /**
+     * Updates existing metadata.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void update(StorageMetadata metadata) throws StorageMetadataException {
+        store.update(this, metadata);
+    }
+
+    /**
+     * Creates a new metadata record.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void create(StorageMetadata metadata) throws StorageMetadataException {
+        store.create(this, metadata);
+    }
+
+    /**
+     * Marks given record as pinned.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void markPinned(StorageMetadata metadata) throws StorageMetadataException {
+        store.markPinned(this, metadata);
+    }
+
+    /**
+     * Deletes a metadata record given the key.
+     *
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void delete(String key) throws StorageMetadataException {
+        store.delete(this, key);
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @throws Exception If transaction could not be commited.
+     */
+    public void commit() throws Exception {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this);
+        isCommitted = true;
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @param lazyWrite true if data can be written lazily.
+     * @throws Exception If transaction could not be commited.
+     */
+    public void commit(boolean lazyWrite) throws Exception {
```

Comment:
In case of crash,  the changes are  lost. However this option should be used only when data can be recreated independently. Eg. length of the chunk is updated lazily. So we don't write to store every time we update chunk length. However in case of failover this field is updated by looking at the actual length of the chunk on the LTS.